build-downstream-rust-repo/action.yaml

```
@@ -0,0 +1,76 @@
+name: Build downstream cargo repo
+description: Build a cargo project that submodules the current repo. Requires pull-request:write permission for PR comments.
+
+inputs:
+    remote_repo:
+      description: The org/repository of the repository to check
+      required: true
+    submodule_path:
+      description: The path in the remote repo of the submodule we are going to try and up-rev
+      required: true
+    build_cmd:
+      description: The command to build the remote repo
+      required: false
+      default: cargo build --release
+    report_success:
+      description: Comment on PR when build is successful.
+      required: false
+      default: 'false'
+
+runs:
+  using: composite
+  steps:
+  - uses: mobilecoinofficial/gh-actions/checkout@v0
+    with:
+      repository: ${{ inputs.remote_repo }}
+
+  - name: Build downstream repo with current branch
+    shell: bash
+    env:
+      SUBMODULE_PATH: ${{ inputs.submodule_path }}
+      BUILD_CMD: ${{ inputs.build_cmd }}
+    run: |
+      set -e
+      set -o pipefail
+
+      # Change to submodule path under remote repo
+      pushd "${SUBMODULE_PATH}"
+
+      # Emulate checkout module and update submodule to current git ref
+      git -c protocol.version=2 fetch --no-tags --prune --no-recurse-submodules --depth=1 origin "+${GITHUB_SHA}:${GITHUB_REF}"
```

Could this be 
```suggestion
      git -c protocol.version=2 fetch --no-tags --prune --no-recurse-submodules --depth=1 origin "+${GITHUB_REF}:${GITHUB_REF}"
```

took a bit to track down, looking at https://docs.github.com/en/actions/learn-github-actions/variables it seems like this will be the SHA of the branch and not of the merge commit, but digging deeper on https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#pull_request
> Note that GITHUB_SHA for this event is the last merge commit of the pull request merge branch. If you want to get the commit ID for the last commit to the head branch of the pull request, use github.event.pull_request.head.sha instead.

---

no change needed, just curious why we fetch by sha instead of by ref, perhaps for concurrent updates/builds

---

I believe this is correct.  I'm looking for the hash of the current commit on the PR.  I tried GITHUB_REF first, but that resolves to something like "refs/remotes/pull/3859/merge"

I did steal this syntax from actions/checkout.  Its fetching the origin <SHA> commit and associating it locally with the value of GITHUB_REF.

I could probably just fetch SHA value instead of doing this mapping, but it feels a little cleaner?

---

I should note that the PR GITHUB_REF ("refs/remotes/pull/3859/merge") isn't directly referenceable after cloning the repo or initializing the submodule.  I think that's a GitHub-thing vs git standard, which is why the fetch and mapping happens.

---

ok that sounds good


> I should note that the PR GITHUB_REF ("refs/remotes/pull/3859/merge") isn't directly referenceable after cloning the repo or initializing the submodule

My understanding, the default git clone will have`+refs/heads/*:refs/remotes/<remote_name>/*` which maps the values on the server's `refs/heads/` into values locally in `refs/remotes/<remote_name>/*`, while `GITHUB_REF` is in `refs/pull` which is *not* part of the common `refs/heads/*` namespace, thus won't be fetched normally by git